### PR TITLE
fix(interpolation): fix bug of interpolation

### PR DESCRIPTION
**Pull request by yhisaki**
## Description

Fixed a bug in the coefficient calculation of spline interpolation.

## Related links

- [TIERIV internal link](https://tier4.atlassian.net/browse/AEAP-1689)

## How was this PR tested?

https://evaluation.tier4.jp/evaluation/reports/1e862479-c0b9-5432-9b06-ca8817bb99f1?project_id=prd_jt

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: common/interpolation/src/spline_interpolation.cpp [hunk 1]
```diff
@@ -166,53 +140,59 @@ std::vector<double> splineByAkima(
...
+    d_ = Eigen::VectorXd::Zero(1);
+    c_[0] = (y[1] - y[0]) / (x[1] - x[0]);
+    d_[0] = y[0];
+    base_keys_ = base_keys;
+    return;
   }
 
+  // Create Tridiagonal matrix
+  Eigen::VectorXd v(n);
+  Eigen::VectorXd h = x.segment(1, n - 1) - x.segment(0, n - 1);
```
Comment (takayuki5168) [inline common/interpolation/src/spline_interpolation.cpp:176]: const

File: common/interpolation/src/spline_interpolation.cpp [hunk 2]
```diff
@@ -18,61 +18,35 @@
...
-      q.push_back((d[i] - a[i - 1] * q[i - 1]) / den);
-    }
+  // Forward sweep
+  c_prime(0) = c(0) / b(0);
+  d_prime(0) = d(0) / b(0);
 
-    // calculate solution
-    x[num_row - 1] = q[num_row - 1];
+  for (auto i = 1; i < n; i++) {
+    double m = 1.0 / (b(i) - a(i - 1) * c_prime(i - 1));
```
Comment (takayuki5168) [inline common/interpolation/src/spline_interpolation.cpp:40]: const

File: common/interpolation/src/spline_interpolation.cpp [hunk 3]
```diff
@@ -18,61 +18,35 @@
...
-    for (size_t i = 1; i < num_row; ++i) {
-      const size_t j = num_row - 1 - i;
-      x[j] = p[j] * x[j + 1] + q[j];
-    }
-  } else {
-    x[0] = (d[0] / b[0]);
+  // Back substitution
+  x(n - 1) = d_prime(n - 1);
+
+  for (auto i = n - 2; i >= 0; i--) {
```
Comment (takayuki5168) [inline common/interpolation/src/spline_interpolation.cpp:48]: I'm concerned about this implementation.
If the type of `i` is size_t (unsigned), substracting 1 from i (= 0) will cause underflow.

File: common/interpolation/src/spline_interpolation.cpp (hunk 1 above)
Comment (yhisaki) [reply common/interpolation/src/spline_interpolation.cpp]: Resolved in https://github.com/autowarefoundation/autoware.universe/pull/8903/commits/55abe390ab4c5b34cb65b2baae199b7715d32a74

File: common/interpolation/src/spline_interpolation.cpp (hunk 2 above)
Comment (yhisaki) [reply common/interpolation/src/spline_interpolation.cpp]: Resolved in https://github.com/autowarefoundation/autoware.universe/pull/8903/commits/55abe390ab4c5b34cb65b2baae199b7715d32a74

File: common/interpolation/src/spline_interpolation.cpp (hunk 3 above)
Comment (yhisaki) [reply common/interpolation/src/spline_interpolation.cpp]: Resolved in https://github.com/autowarefoundation/autoware.universe/pull/8903/commits/e8ce59fbc628929da0ee8da7a0cdaab4e938560c

File: common/interpolation/include/interpolation/spline_interpolation.hpp [hunk 4]
```diff
@@ -98,11 +81,17 @@ class SplineInterpolation
...
+  Eigen::VectorXd c_;
+  Eigen::VectorXd d_;
+
   std::vector<double> base_keys_;
-  interpolation::MultiSplineCoef multi_spline_coef_;
 
   void calcSplineCoefficients(
     const std::vector<double> & base_keys, const std::vector<double> & base_values);
+
+  Eigen::Index get_index(double key) const;
```
Comment (takayuki5168) [inline common/interpolation/include/interpolation/spline_interpolation.hpp:94]: const

**Review by takayuki5168 — APPROVED** [review]

File: common/interpolation/include/interpolation/spline_interpolation.hpp (hunk 4 above)
Comment (yhisaki) [reply common/interpolation/include/interpolation/spline_interpolation.hpp]: Thanks!!
https://github.com/autowarefoundation/autoware.universe/pull/8903/commits/329ac4cd25d0d38eff07fce723314d3727605b48

**Pull request merged**